lib/stellar/sep10.rb

```diff
@@ -0,0 +1,364 @@
+        raise InvalidSep10ChallengeError.new(
+          "The transaction's operation value should be a 64 bytes base64 random string"
+        )
+      end
+
+      if !verify_tx_signed_by(tx_envelope: envelope, keypair: server)
+        raise InvalidSep10ChallengeError.new(
+          "The transaction is not signed by the server"
+        )
+      end
+
+      time_bounds = transaction.time_bounds
+      now = Time.now.to_i
+
+      if time_bounds.nil? || !now.between?(time_bounds.min_time, time_bounds.max_time)
+        raise InvalidSep10ChallengeError.new("The transaction has expired")        
+      end
+
+      # Mirror the return type of the other SDK's and return a string
+      client_kp = Stellar::KeyPair.from_public_key(client_account_id.ed25519!)
+
+      return envelope, client_kp.address
+    end
+
+
+    Contract(C::KeywordArgs[
+      challenge_xdr: String,
+      server: Stellar::KeyPair,
+      signers: SetOf[::Hash]
+    ] => C::SetOf[::Hash])
+    # Verifies that for a SEP 10 challenge transaction all signatures on the transaction are accounted for.
+    #
+    # A transaction is verified if it is signed by the server account, and all other signatures match a signer 
+    # that has been provided as an argument. Additional signers can be provided that do not have a signature, 
+    # but all signatures must be matched to a signer for verification to succeed. 
+    #
+    # If verification succeeds a list of signers that were found is returned, excluding the server account ID.
+    #
+    # @param challenge_xdr [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::Keypair] keypair for server's account.
+    # @param signers [SetOf[::Hash]] The signers of client account.
+    #
+    # @return [SetOf[::Hash]]
+    #
+    # Raises a InvalidSep10ChallengeError if:
+    #     - The transaction is invalid according to Stellar::SEP10.read_challenge_tx.
+    #     - One or more signatures in the transaction are not identifiable as the server account or one of the 
+    #       signers provided in the arguments.
+    def self.verify_challenge_tx_signers(
+      challenge_xdr:,
+      server:,
+      signers:
+    )
+      if signers.empty?
+        raise InvalidSep10ChallengeError.new("No signers provided.")
+      end
+
+      te, _ = read_challenge_tx(
+        challenge_xdr: challenge_xdr, server: server
+      )
+
+      # ignore non-G addresses
+      #
+      # We're also including the server in the list of signers past to verify_tx_signatures
+      # in order to verify that no signature is consumed more than once.
+      g_signers = Set.new
+      signers.each do |signer|
+        begin
+          Stellar::Util::StrKey.check_decode(:account_id, signer['key'])
+        rescue
+          next
+        else
+          g_signers.add(signer)
+        end
+      end
+
+      signers_found = verify_tx_signatures(
+        tx_envelope: te, signers: g_signers
```

Comment:
I don't see the server key being added to the list of signers being passed in here. It's mentioned in the comment a few lines up but I don't see it happening. It's important so that we validate all the signatures in one hit. We shouldn't rely on the caller including the server key in the signers list, although they may.
```suggestion
        tx_envelope: te, signers: g_signers+[server.public_key]
```

You also need code in this function to check that the server was in the list of signers found.

See the [Go SDK](https://github.com/stellar/go/blob/edbd8a0ac1d16206ce22f3ebdf00706b26657b46/txnbuild/transaction.go#L598-L622).

Previously mentioned here: https://github.com/stellar/ruby-stellar-sdk/pull/69#discussion_r392523965.